### PR TITLE
proposal: Live resize vCPU and Memory

### PR DESCRIPTION
**Pull request by kvaps**
related issues:

- https://github.com/kubernetes/kubernetes/pull/102884
- https://github.com/kubevirt/kubevirt/issues/398
- https://github.com/kubevirt/kubevirt/issues/8787

### CODE REVIEW

File: design-proposals/cpu-mem-hotplug.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+- `domain.cpu.sockets` (only if `maxSockets` specified, the value must not be greater than `maxSockets`)
+- `domain.memory.guest` (can be scaled up and scaled down with no troubles)
+- `resources.requests.cpu`
+- `resources.requests.memory`
+- `resources.limits.cpu`
+- `resources.limits.memory`
+
+
+The memory ballooning device should be enabled back to make this feature working.  
+*(it was disabled in v0.59 due to uselessness, related PR: https://github.com/kubevirt/kubevirt/pull/1367)*
```
Comment (olegznv) [inline design-proposals/cpu-mem-hotplug.md:75]: _the_ related

File: design-proposals/cpu-mem-hotplug.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+Author: Andrei Kvapil \<kvapss@gmail.com\>
+
+## Overview
+
+Increase the amount of vCPU and RAM available for the guest OS on the fly.
+
+## Motivation
+
+This is a standard task for any virtualization system.  
+Therefore, virtual machines should support hotplugging the CPU and RAM and be scalable vertically without reboot.
```
Comment (kvaps) [inline design-proposals/cpu-mem-hotplug.md:12]: фывфыв

**Comment by jean-edouard** [comment]
Thank you @kvaps for the proposal! I agree with @xpivarc especially on his first point and would love to see 2 distinct proposals for CPU and memory.

I have started looking into CPU hotplug myself, and the new k8s vertical scaling API is definitely good news for us, but please note that the API will be `alpha` in k8s 1.27. We usually avoid relying on such early APIs because of their volatility, which could mess pretty badly with things like upgrades. ( https://kubernetes.io/docs/reference/using-api/#api-versioning )
Maybe it's worth waiting for at least a beta to start implementing it.

Someone suggested to me that we write the first version of the KubeVirt CPU hotplug feature focusing only on libvirt. Meaning that we will be able to adjust how many vCPUs are visible in a given guest but without actually increasing how much CPU the virt-launcher pod gets. That would make the feature a bit useless, however:
- We could make it so that next time the guest migrates / hard-reboots, the CPU request of the new virt-launcher gets adjusted accordingly
- The effort would be more manageable and the PR(s) smaller
- We will be able to transparently add vertical scalability once the API has matured, without having to change our API at all
I like that idea, even if it sounded a bit crazy at first.

**Comment by kvaps** [comment]
I migrated my fork from `kvaps/community` to `deckhouse/3p-kubevirt-community`, now it will be supported by deckhouse project
fyi

/remove-lifecycle rotten

**Comment by xpivarc** [comment]
/close 
There is already a working [cpu hotplug](https://github.com/kubevirt/community/blob/main/design-proposals/cpu-hotplug.md) and in progress [memory hotplug](https://github.com/kubevirt/community/pull/230)

**Pull request closed without merging**